discojs/src/privacy.ts

```diff
@@ -1,46 +1,39 @@
-import * as tf from '@tensorflow/tfjs'
+import * as tf from "@tensorflow/tfjs";
 
-import type { Task, WeightsContainer } from './index.js'
+import type { WeightsContainer } from "./index.js";
 
-/**
- * Add task-parametrized Gaussian noise to and clip the weights update between the previous and current rounds.
- * The previous round's weights are the last weights pulled from server/peers.
- * The current round's weights are obtained after a single round of training, from the previous round's weights.
- * @param updatedWeights weights from the current round
- * @param staleWeights weights from the previous round
- * @param task the task
- * @returns the noised weights for the current round
- */
-export function addDifferentialPrivacy (updatedWeights: WeightsContainer, staleWeights: WeightsContainer, task: Task): WeightsContainer {
-  const noiseScale = task.trainingInformation?.noiseScale
-  const clippingRadius = task.trainingInformation?.clippingRadius
+async function frobeniusNorm(weights: WeightsContainer): Promise<number> {
+  const squared = await weights
+    .map((w) => w.square().sum())
+    .reduce((a, b) => a.add(b))
+    .data();
+  if (squared.length !== 1) throw new Error("unexcepted weights shape");
 
-  const weightsDiff = updatedWeights.sub(staleWeights)
-  let newWeightsDiff: WeightsContainer
+  return Math.sqrt(squared[0]);
+}
 
-  if (clippingRadius !== undefined) {
-    // Frobenius norm
-    const norm = weightsDiff.frobeniusNorm()
+/** Scramble weights */
+export function addNoise(
+  scale: number,
+  weights: WeightsContainer,
+): WeightsContainer {
+  const deviation = Math.pow(scale, 2);
+  return weights.map((w) => w.add(tf.randomNormal(w.shape, 0, deviation)));
+}
+
+/**
+ * Keep weights' norm within radius
+ *
+ * @param radius maximum norm
+ **/
+export async function clipNorm(
+  radius: number,
+  weights: WeightsContainer,
+): Promise<WeightsContainer> {
```

Comment:
```suggestion
export async function clipNorm(
  weights: WeightsContainer,
  radius: number,
): Promise<WeightsContainer> {
```
There is also tf.clipByValue or x.clipByValue
Nitpicking but I'm used to libraries expecting weights/data as first argument and then the actual parameters (e.g. tf.clipByValue(x, 0, 1))

Reply:
this is clearly more for ML's afficionados: the closer to know libraries, the better!